### PR TITLE
Make Groups  model configurable

### PR DESCRIPTION
**Pull request by bernd-wechner**
Plugging this app into a tendenci site, they do not use auth.Groups but their own user_groups app. To wit made this a configurable option. 

Also refactored the defaults.py as code was mighty confusing named as it was. Regular use of defaults('setting') suggesting the default variable was used when in fact the configured one was taking priority and default filled if none provided. Names are now more indicative of their purpose. I Hope that's pleasing.

All pytests passed. A hopper data fill works fine. And it's plugged into Tendenci using their User Groups, admirably, simply by setting:
```
TODO_USER_GROUP_ATTRIBUTE = "user_groups"
```

it defaults to `groups` and works as it always did, the new option has zero impact on normal use (which took some doing because it turns out Django `ManyToManyRel` has no consistent direction ... but it can be divined.

### CODE REVIEW

File: todo/forms.py [hunk 1]
```diff
@@ -29,8 +33,17 @@ class AddEditTaskForm(ModelForm):
...
     def __init__(self, user, *args, **kwargs):
         super().__init__(*args, **kwargs)
+        
+        # Django's ManyToManyRel is a little odd in that its directional sense 
+        # is not guranteeed and must be tested for. One of these models is User, 
+        # the other is the Groups model (the one group_field points to). We seek 
+        # the attribute with which we can access the set of users in a group.  
+        group_field = getattr(user._meta.model, setting("TODO_USER_GROUP_ATTRIBUTE"), "groups")
+        candidates = (group_field.rel.related_name, group_field.rel.field.attname)
+        user_attr = candidates[1] if group_field.rel.model == user._meta.model else candidates[0]
```
Comment (shacker) [inline todo/forms.py:43]: I can almost but not quite track what this is doing. Can you think of a clearer way to write the python, or add another comment? 

It seems like it is having to recompute for every accessing user every time, whereas it's going to be a project-wide constant for everyone. That seems wasteful. 

Wait - Could this become simpler this way? I think we can assume that any alternate Groups system subclasses django Group. Therefore can't we do something like 

```
if not hasattr(Group, 'parent'):
    setting("TODO_USER_GROUP_ATTRIBUTE")
```

That's not complete but I think you see where I'm going with it - just detect whether the current Groups model has been subclassed.

Man you'd think Django would just have a setting for establishing an alternate Group model. Then it would be easy.

I've thrown out several ideas here, sorry it's not complete, just spitballing clearer and less expensive ways to do this.

File: todo/forms.py [hunk 2]
```diff
@@ -29,8 +33,17 @@ class AddEditTaskForm(ModelForm):
...
+        
         task_list = kwargs.get("initial").get("task_list")
-        members = task_list.group.user_set.all()
+        members = getattr(task_list.group, user_attr).all()
```
Comment (shacker) [inline todo/forms.py:46]: Whew! Funky. Are you confident this will work for any possible alternate group system?

File: todo/management/commands/hopper.py [hunk 3]
```diff
@@ -136,7 +146,16 @@ def add_details(self, build, extracted, **kwargs):
         fake = Faker()  # Use to create user's names
         taskgroup = self.task_list.group
 
-        self.created_by = taskgroup.user_set.all().order_by("?").first()
+        # Django's ManyToManyRel is a little odd in that its directional sense 
+        # is not guranteeed and must be tested for. One of these models is User, 
+        # the other is the Groups model (the one group_field points to). We seek 
+        # the attribute with which we can access the set of users in a group.
+        user_model = get_user_model()
+        group_field = getattr(user_model, setting("TODO_USER_GROUP_ATTRIBUTE"), "groups")
```
Comment (shacker) [inline todo/management/commands/hopper.py:154]: Not DRY

File: todo/models.py [hunk 4]
```diff
@@ -53,9 +55,21 @@ def __enter__(self):
...
+    # The User Group attribute is configurable and so the FreignKey for group
+    # must point to the appropriate model that this attribute relates to.
+    # To wit, TODO_USER_GROUP_ATTRIBUTE must be set before the initial migration
+    # to ensure any dtabase constraints are correctly applied. Djangos migration
+    # tools are not savvy enough to detect changes applied via TODO_USER_GROUP_ATTRIBUTE.  
+    group_field = getattr(User, setting("TODO_USER_GROUP_ATTRIBUTE"), "groups")
+    # Django's ManyToManyRel is a little odd in that its directional sense 
+    # is not guranteeed and must be tested for. One of these models is User, 
+    # the other is the Groups model (the one group_field points to)  
+    candidates = (group_field.rel.model, group_field.rel.related_model)
```
Comment (shacker) [inline todo/models.py:67]: Again

File: todo/settings.py [hunk 5]
```diff
@@ -1,28 +1,29 @@
 # If a documented django-todo option is NOT configured in settings, use these values.
 from django.conf import settings
 
-hash = {
+defaults = {
```
Comment (shacker) [inline todo/settings.py:4]: OK with renaming this hash, but I'm not clear why you renamed it from defaults to settings. The purpose here is to establish a value if the dev didn't set one in settings. This seems to be making it into something other than what it's intended for.

File: todo/management/commands/hopper.py [hunk 6]
```diff
@@ -56,37 +56,47 @@ def handle(self, *args, **options):
 
         fake = Faker()  # Use to create user's names
 
+        user_model = get_user_model()
+        group_field = getattr(user_model, setting("TODO_USER_GROUP_ATTRIBUTE"), "groups")
+        # Django's ManyToManyRel is a little odd in that its directional sense 
+        # is not guranteeed and must be tested for. One of these models is User, 
+        # the other is the Groups model (the one group_field points to)  
+        candidates = (group_field.rel.model, group_field.rel.related_model)
```
Comment (shacker) [inline todo/management/commands/hopper.py:64]: Wait, didn't we just do this in another function? This is not DRY. Find a way to make this (or whatever we decide to go with) reusable (and thus testable).

File: todo/forms.py (hunk 1 above)
Comment (shacker) [reply todo/forms.py:43]: Wait! We don't need to detect `parent` attribute, and we don't need to inspect every users.  All you need to do is to have a shared/reusable function that resolves `TODO_USER_GROUP_ATTRIBUTE` into a real model. That's it - done! Should be just a line or two.

File: todo/tests/test_utils.py [hunk 7]
```diff
@@ -56,24 +56,24 @@ def test_send_email_to_thread_participants(todo_setup, django_user_model, email_
 
 
 def test_defaults(settings):
-    """todo's `defaults` module provides reasonable default values for unspecified settings.
-    If a value is NOT set, it should be pulled from the hash in defaults.py.
+    """todo's `settings` module provides reasonable default values for unspecified settings.
+    If a value is NOT set, it should be pulled from the defaults in settings.py.
     If a value IS set, it should be respected.
-    n.b. TODO_STAFF_ONLY which defaults to True in the `defaults` module."""
+    n.b. TODO_STAFF_ONLY which defaults to True in the `settings` module."""
```
Comment (shacker) [inline todo/tests/test_utils.py:62]: I'm not understanding the point of this change.

**Review by shacker — APPROVED** [review]
This will be a nice addition once we resolve the best way to do it. I have a number of comments and suggestions here, pardon me for thinking aloud. But I think it'll be a lot easier in the end than what you've got here so far. Thanks and hope you don't mind continuing to work on it.

Whatever you do, do it in ONE function somewhere, calling that function wherever needed, and make a test for it please! (if feasible?)

**Review by shacker — CHANGES_REQUESTED** [review]
Sorry I clicked the wrong thing there - that review was meant to be a Request Changes, not Approve.

**Comment by bernd-wechner** [comment]
Thanks kindly for the review. I shall respond and together I hope we can make it satisfactory. It's running fine now as is on our live site, but of course to mainline this flexibility it needs to work generically and your review is gratefully accepted.

File: todo/forms.py [hunk 8]
```diff
@@ -29,8 +33,17 @@ class AddEditTaskForm(ModelForm):
 
     def __init__(self, user, *args, **kwargs):
         super().__init__(*args, **kwargs)
+        
+        # Django's ManyToManyRel is a little odd in that its directional sense 
+        # is not guranteeed and must be tested for. One of these models is User, 
+        # the other is the Groups model (the one group_field points to). We seek 
+        # the attribute with which we can access the set of users in a group.  
+        group_field = getattr(user._meta.model, setting("TODO_USER_GROUP_ATTRIBUTE"), "groups")
```
[the remark this answers is not in the document]
Comment (bernd-wechner) [reply todo/forms.py:41]: `group_field` is used as a name twice in two contexts to mean different things. The one above is the field on the `user` object that tells use what groups the user is in. The second is a field on the `User` model which stores the relation to the `Group` model. They are different types of object. There's an argument to name them differently, and I'm open to that. But this second one is just a intermediary for getting the two related field names. One of which is `User` and the other of which is the `Group`. Alas all written in generics. That is `User` is `user._meta.model` and `Group` is one of `group_field.rel.related_name` and `group_field.rel.field.attname`.

I'll take a double look at it again (it took quite a while to get this write as this inconsistent directivity of the Django Relation is a bother. I suspect this was the simplest approach but will double check and report.

File: todo/forms.py (hunk 1 above)
Comment (bernd-wechner) [reply todo/forms.py:43]: All good. Indeed, this could be a generic Django utility function (should really be in Django models already. It's asking "What is the related model for this field?" and because the field is a Django relation it turns out that I can't find that answer in one place. There are two models stored in oddly inconsistent places in the relation object and which one is this model and which one is the related model seems to probably depend on which model was declared first at a guess! Not 100% sure, but it's not reliable)

File: todo/forms.py (hunk 2 above)
Comment (bernd-wechner) [reply todo/forms.py:46]: Pretty confident. As the only premise is that there is a ManyToMany relation between the group model and `User` (which is fairly safe, but to be honest as `User` is a fairly fixed entity in auth (can be overridden too but I'd guess that unlikely as Django explicitly provides a profile paradigm for extending it which is commonly used) and any new user group model must have a ManyToMany to `User` or a ForeignKey to `User` but either way that is a ToMany relation from User and hence the related_name that an admin must specify in `TODO_USER_GROUP_ATTRIBUTE` is a ToMany relation, hence supporting .all().

To be honest though as this requires setting `TODO_USER_GROUP_ATTRIBUTE` I epxect an admin to be on top of that ;-). To be perfectly safe, you cover the unlikely case that an admin specifies a OneToOne relation between User and their User Groups ... in that instance and that one alone, it warrants a test, to see if `.all()` is implemented to return the one related object (for consistency) or of Django implements another way of getting it.

If the latter, we'd just need to pull the same trick, `getattr` and if `all` exists check if its callable.

File: todo/management/commands/hopper.py (hunk 6 above)
Comment (bernd-wechner) [reply todo/management/commands/hopper.py:64]: True, this can go to a small utility function.

File: todo/management/commands/hopper.py (hunk 3 above)
Comment (bernd-wechner) [reply todo/management/commands/hopper.py:154]: Again, can go to a utility function. Will do that on all these repeats. Agreed. DRY is better.

File: todo/settings.py (hunk 5 above)
Comment (bernd-wechner) [reply todo/settings.py:4]: Mainly because elsewhere in the code you were asking for `defaults('setting_name')` to get the value of a setting. You're not getting the default at all. You're feting the value of the setting, unless it is not set then returning a default. This is in fact exactly what the rest of Django does, settings all have default values. But to get the value of the setting `setting('setting_name')` is much clearer - it says what you are fetching.

File: todo/tests/test_utils.py (hunk 7 above)
Comment (bernd-wechner) [reply todo/tests/test_utils.py:62]: Explained above.

File: todo/forms.py (hunk 1 above)
Comment (shacker) [reply todo/forms.py:43]: Would you mind if I gave it a try?

File: todo/settings.py (hunk 5 above)
Comment (shacker) [reply todo/settings.py:4]: Hmm, I do see your point. On the other hand, we have a convention that `settings.py` is where you put your settings, and that is not what this file is for! By giving it the same filename, it implies that it's for storing settings.

File: todo/forms.py (hunk 1 above)
Comment (bernd-wechner) [reply todo/forms.py:43]: Not sure what you mean, but no I don't mind ...

File: todo/settings.py (hunk 5 above)
Comment (bernd-wechner) [reply todo/settings.py:4]: That is a minor concern to me, as `settings.py` for conf settings is generally very well identifies. `defaults.py` I find wholly inappropriate though. Perhaps moving that small utility into a utility file gets around the whole naming issue. it's a tiny thing. In fact it's not entirely out of place in `settings.py` itself IMHO.

**Comment by shacker** [comment]
@bernd-wechner I got really close to a working solution, but then I realized something: If you define a custom groups model, it will *always* have to inherit from Django `Group`:

```
class MyGroup(Group):
   xxx
```

and if users are added to that group, and you then call `user.groups.all()`, it just works - you get a queryset of groups that are  instances of the custom group. So now I'm confused - after all of this, why is this ticket needed at all? In what case does `user.groups.all()` or `.filter()` not work? 

If the custom group does not inherit from Django `Group`, then that is a bug in the project site, not a problem to be solved in django-todo! Right?

**Comment by bernd-wechner** [comment]
> @bernd-wechner I got really close to a working solution, but then I realized something: If you define a custom groups model, it will _always_ have to inherit from Django `Group`:
> 
> ```
> class MyGroup(Group):
>    xxx
> ```
> 
> and if users are added to that group, and you then call `user.groups.all()`, it just works - you get a queryset of groups that are instances of the custom group. So now I'm confused - after all of this, why is this ticket needed at all? In what case does `user.groups.all()` or `.filter()` not work?
> 
> If the custom group does not inherit from Django `Group`, then that is a bug in the project site, not a problem to be solved in django-todo! Right?

Wow, that's news. It was crashing. I will definitely try that and report back. I'm snowed under right now on another job. But soon.

**Comment by shacker** [comment]
I hear you! I've spent two evenings on this PR and don't really have time to. It's interesting for sure, but if the problem is already a solved one in Django, all the better. Let me know what you figure out.

**Comment by bernd-wechner** [comment]
> I hear you! I've spent two evenings on this PR and don't really have time to. It's interesting for sure, but if the problem is already a solved one in Django, all the better. Let me know what you figure out.

Well, I'm doing some last minute tidyp before a conference that the website will need to up for. And it has the TOdos. So Ic hecked and no, your premise is wrong:

```
class Group(TendenciBaseModel):
```

is what Tendenci does. And they may be odd, but it cannot be unique. In short User group management can be reimplemented (lord knows why you would but that is not mine to ponder, I need this running on Tendenci sites, and the PR enables it - so I'm thinking I tidy it up DRY and it'll be a winner!). That will wait till next week at earliest.

**Comment by shacker** [comment]
> class Group(TendenciBaseModel):

Right, the question is, is Tendenci doing it wrong? Just because it's possible doesn't mean it's right! I admit that my statement was based on the fact that I've never *seen* anyone create a custom group model that did not inherit from Django Groups. That doesn't mean it's "illegal." I've posted this [question](https://groups.google.com/g/django-users/c/UnR5IUqA0bM) on the Django-users mailing list - hoping for some insight. 

One option would be to simply document in Django-todo: "Not compatible with projects that use a groups system that does not subclass the Django Group model." Why should we bend over backwards to support apps that are doing it wrong? But I'll wait to hear feedback before doing that.

**Comment by bernd-wechner** [comment]
"wrong" is so subjective it's not really relevant IMHO. It's one datum that suggests it's booth possible and done (at least once, and any social demographer would project from that a loose forecast that it's at least now and again - i.e. the probability of having found a unique exemplar is eclipsed by the probability that it is not a unique exemplar) and that I certainly need it working in a  Tendenci framework.  

For whatever reason they did no extend auth.Group, and the reasons are probably lost in time, though you have made me curious. I wonder if anyone at Tendenci knows? It's more a curio though than profoundly important because it falsifies the hypothesis that user group management **always has to**  inherit from auth.Group - because Tendenci User Group Management works just fine (even if it is heretic under the hood and reimplements its own Group model.

**Comment by shacker** [comment]
I'd say it's subjective if the framework has no stated position or best practices on the matter. But if they have documentation where they recommend it be done in a particular way, and someone chooses to ignore that, it's a different matter. A framework works if all parts of the ecosystem are designed for interoperability. Example: In your templates you can call your blocks anything you like. But if you release code for others to use, you damn well better use standard block names like `{% block content %}` or things just won't work together. If you made a PR that asked me to support a different main block name in the `todo` templates, I would reject it. 

No answer from the Django-users mailing list today. I'll ask again on Django-developers and see if anything comes up.

If you have contact with anyone at Tendenci and can pitch the same question, I'd be curious what they have to say.

**Comment by shacker** [comment]
@bernd-wechner OK we finally got a response from one of the Django core maintainers on this. Aymeric says 

"The semi-official position is "create a separate model with a FK to the auth.Group". I know that's less than helpful. At least you know that there's nothing better built-in." ([link](https://groups.google.com/g/django-developers/c/JEAsJbIrLJM))

That's pretty much what I expected. Technically a gray area since it's not documented, but there is a semi-official position anyway. 

So here's my take: If you can find a way to make this elegant, non-intrusive, and performant (i.e. doesn't cost an extra query on every request just to figure this out), and compacted into a re-usable function, I guess I'm open to it. But the *best* solution would be to try and figure out what Tendenci was thinking when they did this, and if they're open to changing it. I just hate to have todo accumulate workarounds for what I consider to be highly unusual edge cases.

Let me know what you think.

Scot

**Comment by bernd-wechner** [comment]
Significant rehash targeting:

1. DRYness (even DRYed up a few other bits ;-)
2. Passes all Tests
3. Added a test for hopper (it's a bit slow but was vital for confirming code integrity as hopper has special needs)
4. Centralising the whole User/Groups thing in defaults.py
5. Renaming defaults.py back to just that, but conserving the function name (settings) and adding the support functions for generalising.
6. Fully tested on a tendenci site, and hopper too and all functional.
7. Passed code inspection my side inspecting those relationship partners
8. Simplified the code of choosing one of the two relationship partners immensely and made it consistently in two contexts.

As to edge cases, I can't argue what is and isn't but I do hear you concern, and take it on board and agree. But that said, I see unrelated issues justifying this (unrelated that is to assessed popularity of different user group modelling):

1. One way or another I need it as does any tendenci app wanting to plug in this app.
2. Maintaining a general position on user management is worthwhile as illustrated by Django's own` auth.get_user_model()` which you lean on. That is Django is cognizant of the fact that a given instance may implement a different user model and so instead of referencing `auth.User` we reference `get_user_model()`. I have analogously implemented `get_group_model()`.

Hopefully it's a pleasing rehash. Am as ever always open to feedback, but would much rather we maintained a portable `django-todo` rather than my maintaining a tendenci compatible form independently.

**Comment by shacker** [comment]
@bernd-wechner This is interesting - I went to the Django bug tracker and searched for `AUTH_GROUP_MODEL` and found [one ticket](https://code.djangoproject.com/ticket/29748) on this topic (there could be more, but this is the only one that came up on that search). It's 3 years old and is in state "Maybe / Someday" i.e. definitely not high priority. There is no commentary from core devs. I searched for a while for related tickets but couldn't find anything. 

Here is my takeaway:

- This issue affects **very** few projects. It's probably fairly uncommon to use a custom Groups model at all, and those who do almost always subclass Group, so things just work. 
- Unfortunately, that technique appears to be undocumented in Django, leading to the misunderstanding we have here. I'm going to ask on the developer list whether they would accept a documentation PR that would solidify the practice.
- If we patch this in Django-todo, it could just happen again in the next reusable app you install. That's just playing whack-a-mole. In other words, I am increasingly feeling like todo is the wrong place to address this. The problem is in Tendenci and in Django's lack of a clear statement on the matter. 

As much as I'd love to accommodate you, and I appreciate your effort on this, it still feels like a fair bit of code churn that ultimately is just for you and maybe other Tendenci users and that you should be submitting PRs to their project instead.

**Comment by shacker** [comment]
@bernd-wechner I kind of dropped the ball on this conversation, apologies. Thanks for your thoughtful responses and the time you put into the PR, but after re-reading the thread above, I remain convinced that this is simply a matter of "Tendenci is doing it wrong and it's not our job to work around that fact." So I have decided to close this without merging. Feel free to post any updates here wrt communications you've gotten from Tendenci on this topic.
